### ISSUE TITLE
Decoding hangs when using LM rescoring

### ISSUE THREAD

**Issue by boliangz** [wenet-e2e/wenet#563]
I'm following this tutorial to use LM rescoring for decoding: https://github.com/wenet-e2e/wenet/blob/23a61b212bf2c3886546925913f5574f779f474a/examples/librispeech/s0/run.sh#L234

I didn't re-train a model and instead, I use the pre-trained conformer model. I had no problem building the TLG.fst, but `./tools/decode.sh` hangs forever when evaluating on the test set. Could you provide any suggestions on where the problem would be and how to debug this?

The below is the code I used for LM rescoring (I took this out from run.sh):
```
pretrained_model=wenet/models/20210216_conformer_exp
dict=$pretrained_model/words.txt
bpemodel=$pretrained_model/train_960_unigram5000

lm=data/local/lm
lexicon=data/local/dict/lexicon.txt
mkdir -p $lm
mkdir -p data/local/dict

# 7.1 Download & format LM
which_lm=3-gram.pruned.1e-7.arpa.gz
if [ ! -e ${lm}/${which_lm} ]; then
    wget http://www.openslr.org/resources/11/${which_lm} -P ${lm}
fi
echo "unzip lm($which_lm)..."
gunzip -k ${lm}/${which_lm} -c > ${lm}/lm.arpa
echo "Lm saved as ${lm}/lm.arpa"

# 7.2 Prepare dict
unit_file=$dict
bpemodel=$bpemodel
# use $dir/words.txt (unit_file) and $dir/train_960_unigram5000 (bpemodel)
# if you download pretrained librispeech conformer model
cp $unit_file data/local/dict/units.txt
if [ ! -e ${lm}/librispeech-lexicon.txt ]; then
    wget http://www.openslr.org/resources/11/librispeech-lexicon.txt -P ${lm}
fi
echo "build lexicon..."
tools/fst/prepare_dict.py $unit_file ${lm}/librispeech-lexicon.txt \
    $lexicon $bpemodel.model
echo "lexicon saved as '$lexicon'"

# 7.3 Build decoding TLG
tools/fst/compile_lexicon_token_fst.sh \
   data/local/dict data/local/tmp data/local/lang
tools/fst/make_tlg.sh data/local/lm data/local/lang data/lang_test || exit 1;

# 7.4 Decoding with runtime
echo "Start decoding..."
fst_dir=data/lang_test
dir=$pretrained_model
recog_set="test_clean"
for test in ${recog_set}; do
    ./tools/decode.sh --nj 2 \
        --beam 10.0 --lattice_beam 5 --max_active 7000 --blank_skip_thresh 0.98 \
        --ctc_weight 0.5 --rescoring_weight 1.0 --acoustic_scale 1.2 \
        --fst_path $fst_dir/TLG.fst \
        data/$test/wav.scp.10 data/$test/text.10 $dir/final.zip $fst_dir/words.txt \
        $dir/lm_with_runtime_${test}
    tail $dir/lm_with_runtime_${test}/wer
done
```

**Comment by robin1001** [comment]
You can add some print lines in decode_main.cc, then rebuild, and see where it hangs.

**Comment by xingchensong** [comment]
hi, can u provide more info such as log file (located in `exp/20210216_conformer_exp/lm_with_runtime_test_clean/split2/x.log`) ? 

BTW,  whats the version (or commit hash) of ur wenet runtime ?  I guess it may be caused by the mismatch between runtime code (dated after 20210601)  and pretrained model (dated in 20210216). I have tried building runtime with the latest code and inferencing with the latest model ([20210610_conformer](http://mobvoi-speech-public.ufile.ucloud.cn/public/wenet/librispeech/20210610_conformer_bidecoder_exp.tar.gz)), it's ok to get correct predictions.

**Comment by gstswwx** [comment]
@boliangz @xingchensong
I trained the model by myself and did not use the pretrained model. The version of torch and torchaudio is 1.9.0 and the version of CUDA is 11.1. The code repository was cloned 1 week ago.

When I run `bash run.sh --stage 7 --stop-stage 7`, I found the program built TLG.fst successfully but failed when executing
`"model->Read(FLAGS_model_path, FLAGS_num_threads);"` in `wenet/runtime/server/x86/decoder/params.h`. It seems that the program failed to load jit model. A screenshot of the errors is shown below.

![截屏2021-08-30 上午10 20 15](https://user-images.githubusercontent.com/21003094/131276322-3a13fdf2-9490-4e4f-ae57-1935075679b5.png)

**Comment by xingchensong** [comment]
> @boliangz @xingchensong
> I trained the model by myself on RTX3090 and did not use the pretrained model. The version of torch and torchaudio is `1.9.0` and `0.9.0`. The version of CUDA is `11.1`. The code repository was cloned 1 week ago.
> 
> When I ran `bash run.sh --stage 7 --stop-stage 7`, I found the program built TLG.fst successfully but failed when executing
> `model->Read(FLAGS_model_path, FLAGS_num_threads);` in `wenet/runtime/server/x86/decoder/params.h`. The model file `final.zip` exsited in the right path. So, it seems that the program failed to load jit model. A screenshot of the errors is shown below.
> 
> ![截屏2021-08-30 上午10 20 15](https://user-images.githubusercontent.com/21003094/131276322-3a13fdf2-9490-4e4f-ae57-1935075679b5.png)

Have u run stage 6 to get correct  `final.zip` ? Also, plz show us your log info (`exp/your_model_name/lm_with_runtime_test_clean/splitX/x.log`)

**Comment by boliangz** [comment]
> hi, can u provide more info such as log file (located in `exp/20210216_conformer_exp/lm_with_runtime_test_clean/split2/x.log`) ?
> 
> BTW, whats the version (or commit hash) of ur wenet runtime ? I guess it may be caused by the mismatch between runtime code (dated after 20210601) and pretrained model (dated in 20210216). I have tried building runtime with the latest code and inferencing with the latest model ([20210610_conformer](http://mobvoi-speech-public.ufile.ucloud.cn/public/wenet/librispeech/20210610_conformer_bidecoder_exp.tar.gz)), it's ok to get correct predictions.

Here is the log:
```
I0816 15:58:53.083853  2498 params.h:83] Reading model /home/boliangzhang/stream_speech_translation/backend/src/services/asr/wenet/models/20210216_conformer_exp/final.zip
I0816 15:58:55.324354  2498 torch_asr_model.cc:38] torch model info subsampling_rate 4 right context 6 sos 5001 eos 5001 is bidirectional decoder 0 num threads 1
I0816 15:58:55.324396  2498 params.h:90] Reading fst data/lang_test/TLG.fst
I0816 15:58:55.324596  2498 fst.h:799] FstImpl::ReadHeader: source: data/lang_test/TLG.fst, fst_type: vector, arc_type: standard, version: 2, flags: 0
I0816 15:58:59.563489  2498 params.h:96] Reading symbol table data/lang_test/words.txt
```

**Comment by robin1001** [comment]
We will upgrade libtorch to the latest version(1.9.0).

**Comment by xingchensong** [comment]
> > hi, can u provide more info such as log file (located in `exp/20210216_conformer_exp/lm_with_runtime_test_clean/split2/x.log`) ?
> > BTW, whats the version (or commit hash) of ur wenet runtime ? I guess it may be caused by the mismatch between runtime code (dated after 20210601) and pretrained model (dated in 20210216). I have tried building runtime with the latest code and inferencing with the latest model ([20210610_conformer](http://mobvoi-speech-public.ufile.ucloud.cn/public/wenet/librispeech/20210610_conformer_bidecoder_exp.tar.gz)), it's ok to get correct predictions.
> 
> I'm on commit `1eec95b373e22e9a0c6686dd87636d47ad04ab3f`.
> 
> Here is the log:
> 
> ```
> I0816 15:58:53.083853  2498 params.h:83] Reading model /home/boliangzhang/stream_speech_translation/backend/src/services/asr/wenet/models/20210216_conformer_exp/final.zip
> I0816 15:58:55.324354  2498 torch_asr_model.cc:38] torch model info subsampling_rate 4 right context 6 sos 5001 eos 5001 is bidirectional decoder 0 num threads 1
> I0816 15:58:55.324396  2498 params.h:90] Reading fst data/lang_test/TLG.fst
> I0816 15:58:55.324596  2498 fst.h:799] FstImpl::ReadHeader: source: data/lang_test/TLG.fst, fst_type: vector, arc_type: standard, version: 2, flags: 0
> I0816 15:58:59.563489  2498 params.h:96] Reading symbol table data/lang_test/words.txt
> ```

Hi, sorry for getting back to you late. I think the type of your wav_file must be *.flac, you should first convert *.flac to *.wav

**Comment by chiendb97** [comment]
> 
When i convert the file format from flac to wav the decoding doesn't hang
However decoding results in wer 100%
Here is the wav file information:
`Channels       : 1`
`Sample Rate    : 16000`
`Precision      : 16-bit`
`Duration       : 00:00:10.44 = 166960 samples ~ 782.625 CDDA sectors`
`File Size      : 334k`
`Bit Rate       : 256k`
`Sample Encoding: 16-bit Signed Integer PCM`
Here is the result:
`WER: 100.00 % N=28 C=0 S=27 D=1 I=0`
`lab: HE HOPED THERE WOULD BE STEW FOR DINNER TURNIPS AND CARROTS AND BRUISED POTATOES AND FAT MUTTON PIECES TO BE LADLED OUT IN  THICK PEPPERED FLOUR FATTENED SAUCE `
`rec: H  HR    L     N     U  N    U   N      N       U   N       UI  YT      N        N   U   N      ZE     L  U  N      HC  ZEP L     U        N     UE  `

**Comment by xingchensong** [comment]
whats ur dict? It should be `data/lang_test/words.txt` if u use wfst otherwise `$model_dir/words.txt`

**Comment by chiendb97** [comment]
> 

I use wfst and dict is data/lang_test/words.txt

**Comment by iPwnXX** [comment]
> > > hi, can u provide more info such as log file (located in `exp/20210216_conformer_exp/lm_with_runtime_test_clean/split2/x.log`) ?
> > > BTW, whats the version (or commit hash) of ur wenet runtime ? I guess it may be caused by the mismatch between runtime code (dated after 20210601) and pretrained model (dated in 20210216). I have tried building runtime with the latest code and inferencing with the latest model ([20210610_conformer](http://mobvoi-speech-public.ufile.ucloud.cn/public/wenet/librispeech/20210610_conformer_bidecoder_exp.tar.gz)), it's ok to get correct predictions.
> > 
> > 
> > I'm on commit `1eec95b373e22e9a0c6686dd87636d47ad04ab3f`.
> > Here is the log:
> > ```
> > I0816 15:58:53.083853  2498 params.h:83] Reading model /home/boliangzhang/stream_speech_translation/backend/src/services/asr/wenet/models/20210216_conformer_exp/final.zip
> > I0816 15:58:55.324354  2498 torch_asr_model.cc:38] torch model info subsampling_rate 4 right context 6 sos 5001 eos 5001 is bidirectional decoder 0 num threads 1
> > I0816 15:58:55.324396  2498 params.h:90] Reading fst data/lang_test/TLG.fst
> > I0816 15:58:55.324596  2498 fst.h:799] FstImpl::ReadHeader: source: data/lang_test/TLG.fst, fst_type: vector, arc_type: standard, version: 2, flags: 0
> > I0816 15:58:59.563489  2498 params.h:96] Reading symbol table data/lang_test/words.txt
> > ```
> 
> Hi, sorry for getting back to you late. I think the type of your wav_file must be *.flac, you should first convert *.flac to *.wav

I met the same issue and located the bug in `decoder_main.cc` wav_reader.  does this librispeech recipe script only support wav format files?  the raw librispeech corpus is in flac format, and i didn't find any readme info or comments suggesting the format conversion.
